File: src/db/repos/orders_repo.py
"""
Repository layer for the `orders` table.

Isolates all data-access logic so the API layer stays thin.
Uses an in-memory store by default; swap `_store` for a real DB session
in production without changing any API code.
"""

from __future__ import annotations

import uuid
from typing import Dict, Optional

from src.db.models import Order


class OrdersRepository:
    """
    CRUD operations against `orders`.

    The `_store` class variable acts as the persistence layer during tests
    and local development.  In production this class would be instantiated
    with a real DB session/connection.
    """

    # In-memory backing store keyed by order_id.
    # Replace with a real DB dependency (e.g. SQLAlchemy session) for production.
    _store: Dict[str, Order] = {}

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get(self, order_id: str) -> Optional[Order]:
        """Return the Order for *order_id*, or None if not found."""
        return self._store.get(order_id)

    def create_order(self, product_id: str, quantity: int, price: str) -> Order:
        """
        Persist a new order and return the created record.

        Parameters
        ----------
        product_id : str
            UUID string identifying the product.
        quantity : int
            Number of units ordered (caller must ensure > 0).
        price : str
            Unit price as a decimal string (caller must ensure >= 0.01).

        Returns
        -------
        Order
            The newly created order record.
        """
        order_id = str(uuid.uuid4())
        order = Order(
            order_id=order_id,
            product_id=product_id,
            quantity=quantity,
            price=price,
        )
        self._store[order_id] = order
        return order

    # ------------------------------------------------------------------
    # Test / seed helpers (not part of the production interface)
    # ------------------------------------------------------------------

    def _seed(self, order: Order) -> None:
        """Insert or overwrite a record — used by tests only."""
        self._store[order.order_id] = order

    def _clear(self) -> None:
        """Remove all records — used by tests only."""
        self._store.clear()

File: go/internal/db/orders_repo.go
package db

import (
	"crypto/rand"
	"fmt"
	"sync"
	"time"

	"github.com/MinnMinn/skills-introduction-to-github/internal/models"
)

// OrdersRepository performs CRUD operations against `orders`.
//
// Uses an in-memory store by default; swap store for a real DB session in
// production without changing any API code.
type OrdersRepository struct {
	mu    sync.RWMutex
	store map[string]*models.Order
}

// NewOrdersRepository returns an initialised, empty repository.
func NewOrdersRepository() *OrdersRepository {
	return &OrdersRepository{
		store: make(map[string]*models.Order),
	}
}

// Get returns the Order for orderID, or nil if not found.
func (r *OrdersRepository) Get(orderID string) *models.Order {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.store[orderID]
}

// CreateOrder persists a new order and returns the created record.
//
// Parameters:
//
//	productID — UUID string identifying the product
//	quantity  — number of units ordered (caller must ensure > 0)
//	price     — unit price as a decimal string (caller must ensure >= 0.01)
func (r *OrdersRepository) CreateOrder(productID string, quantity int, price string) (*models.Order, error) {
	orderID, err := newUUID()
	if err != nil {
		return nil, fmt.Errorf("generate order_id: %w", err)
	}

	order := &models.Order{
		OrderID:   orderID,
		ProductID: productID,
		Quantity:  quantity,
		Price:     price,
		Status:    "pending",
		CreatedAt: time.Now().UTC().Format(time.RFC3339Nano),
	}

	r.mu.Lock()
	defer r.mu.Unlock()
	r.store[orderID] = order
	return order, nil
}

// newUUID generates a random UUID v4 string using only the standard library.
func newUUID() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", err
	}
	// Set version 4 and variant bits
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf(
		"%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:],
	), nil
}

// ---------------------------------------------------------------------------
// Test / seed helpers (not part of the production interface)
// ---------------------------------------------------------------------------

// Seed inserts or overwrites a record — used by tests only.
func (r *OrdersRepository) Seed(o *models.Order) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.store[o.OrderID] = o
}

// Clear removes all records — used by tests only.
func (r *OrdersRepository) Clear() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.store = make(map[string]*models.Order)
}
